codegen/resolver.go

```diff
@@ -0,0 +1,78 @@
+package codegen
+
+import (
+	"context"
+	"sync"
+
+	"github.com/docker/buildx/util/progress"
+	"github.com/moby/buildkit/client"
+	"github.com/moby/buildkit/client/llb"
+	gateway "github.com/moby/buildkit/frontend/gateway/client"
+	digest "github.com/opencontainers/go-digest"
+	"github.com/openllb/hlb/pkg/llbutil"
+	"github.com/openllb/hlb/solver"
+	"golang.org/x/sync/errgroup"
+)
+
+func NewCachedImageResolver(cln *client.Client) llb.ImageMetaResolver {
+	return &cachedImageResolver{
+		cln:   cln,
+		cache: make(map[string]*imageConfig),
+	}
+}
+
+type cachedImageResolver struct {
+	cln   *client.Client
+	cache map[string]*imageConfig
+	mu    sync.RWMutex
+}
+
+type imageConfig struct {
+	dgst   digest.Digest
+	config []byte
+}
+
+func (r *cachedImageResolver) ResolveImageConfig(ctx context.Context, ref string, opt llb.ResolveImageConfigOpt) (dgst digest.Digest, config []byte, err error) {
+	r.mu.RLock()
+	cfg, ok := r.cache[ref]
+	if ok {
+		r.mu.RUnlock()
```

Comment:
I would move the `r.mu.RUnlock()` before the `if ok` check so you can dedup and remove the one on line 42